src/js/parsers/dom.js

```diff
@@ -89,6 +101,33 @@ function walkMarkerableNodes(parent, callback) {
   }
 }
 
+function walkSectionNodes(parent, callback) {
+  var currentNode = parent;
+
+  if (isTextNode(currentNode)) {
+    // If we found a text node with no siblings, we probably went too far
+    if (currentNode.parentNode.childNodes.length === 1) {
+    callback(currentNode.parentNode);
+    } else {
+      callback(currentNode);
+    }
+
+  // If we find a valid section, callback
+  } else if (isValidSection(currentNode)) {
+    callback(currentNode);
+  // Otherwise, look at the next available node
+  } else {
+    if (!currentNode.firstChild) {
+      callback(currentNode);
+    }
+    currentNode = currentNode.firstChild;
+    while (currentNode) {
+      walkSectionNodes(currentNode, callback);
```

Comment:
It would be great to have the recursion flattened out of this function. Instead using a while loop and current node to descend and ascend in a single function. Mostly a style issue, sometimes a performance issue but I doubt so in this case.

Reply:
Could you expand a little on what you mean here?